File: paleo_workbench/ui/pages/sequence_framework_page.py
from __future__ import annotations

from PySide6.QtCore import QEvent, Qt, QTimer, Signal
from PySide6.QtWidgets import QMessageBox, QSplitter, QToolButton, QVBoxLayout, QWidget

from paleo_workbench.ui import tokens
from paleo_workbench.ui.dock_manager import dock_manager
from paleo_workbench.ui.layout_persistence import LayoutPersistence
from paleo_workbench.ui.panel_float_controller import FloatController
from paleo_workbench.ui.pages.sequence_boundary_table import SequenceBoundaryTable
from paleo_workbench.ui.pages.sequence_scheme_summary import SequenceSchemeSummary
from paleo_workbench.ui.pages.sequence_target_panel import SequenceTargetPanel
from paleo_workbench.workflow.stratigraphy import (
    apply_stratigraphy_scheme,
    set_target_from_boundary,
)

# QWIDGETSIZE_MAX: lifts a side panel's setFixedWidth upper bound (the
# designed width stays as the minimum) so the splitter handle can resize it.
_PANEL_MAX_WIDTH = 16_777_215

#: Delay before a splitter drag is persisted (avoid a QSettings sync per tick).
_DOCKED_SIZES_DELAY_MS = 400


class PanelFloatButton(QToolButton):
    """Corner button floating one side panel via a FloatController.

    Docked side of the FloatingPanel ⇲ dock-back chrome: pinned to the
    panel's top-right corner through an event filter and hidden while the
    panel is afloat (the floating window carries its own dock-back button).
    """

    def __init__(self, key: str, panel: QWidget, controller: FloatController):
        super().__init__(panel)
        self._key = key
        self._panel = panel
        self._controller = controller
        self.setObjectName("PanelFloatButton")
        self.setText("⇱")
        self.setToolTip("浮动面板 (Float panel)")
        self.setFixedSize(18, 18)
        self.clicked.connect(lambda: self._controller.toggle(self._key))
        panel.installEventFilter(self)
        controller.float_changed.connect(self._on_float_changed)
        self._reposition()

    def _reposition(self) -> None:
        self.move(self._panel.width() - self.width() - 4, 2)

    def eventFilter(self, obj, event) -> bool:  # noqa: N802 (Qt override)
        if obj is self._panel and event.type() == QEvent.Type.Resize:
            self._reposition()
        return False

    def _on_float_changed(self, key: str, floating: bool) -> None:
        if key == self._key:
            self.setVisible(not floating)


class SequenceFrameworkPage(QWidget):
    """层序格架 page: edit scheme, persist stratigraphy, bind target_horizon downstream."""

    stratigraphy_updated = Signal()

    def __init__(self, parent=None, *, persistence: LayoutPersistence | None = None):
        super().__init__(parent)
        self.setObjectName("SequenceFrameworkPage")
        self._project = None

        outer = QVBoxLayout(self)
        outer.setContentsMargins(
            tokens.PAGE_MARGIN,
            tokens.PAGE_MARGIN,
            tokens.PAGE_MARGIN,
            tokens.PAGE_MARGIN,
        )
        outer.setSpacing(tokens.SPACE_4)

        self.content_splitter = QSplitter(Qt.Orientation.Horizontal)
        self.content_splitter.setObjectName("SequenceFrameworkSplitter")
        self.content_splitter.setChildrenCollapsible(False)

        self.target_panel = SequenceTargetPanel()
        # 目标 | 层序界面表 | 方案: the boundary table stays the stretchy
        # center; target + summary are resizable side panels.
        self.target_panel.setMaximumWidth(_PANEL_MAX_WIDTH)
        self.content_splitter.addWidget(self.target_panel)

        self.boundary_table = SequenceBoundaryTable()
        self.content_splitter.addWidget(self.boundary_table)

        self.scheme_summary = SequenceSchemeSummary()
        self.scheme_summary.setMaximumWidth(_PANEL_MAX_WIDTH)
        self.content_splitter.addWidget(self.scheme_summary)

        self.content_splitter.setStretchFactor(0, 0)
        self.content_splitter.setStretchFactor(1, 1)
        self.content_splitter.setStretchFactor(2, 0)
        self.content_splitter.setSizes([280, 900, 260])

        outer.addWidget(self.content_splitter, 1)

        # M6: target/scheme side panels float through the shared
        # FloatController; the boundary table stays the docked center.
        self._float_persistence = (
            persistence if persistence is not None else LayoutPersistence()
        )
        self._floatable: dict[str, QWidget] = {}
        self.float_controller = FloatController(
            resolver=self._floatable.get,
            persistence=self._float_persistence,
            parent=self,
        )
        self._make_floatable("sequence:target", self.target_panel, "层序目标")
        self._make_floatable("sequence:scheme", self.scheme_summary, "层序方案")
        self._float_sizes_timer = QTimer(self)
        self._float_sizes_timer.setSingleShot(True)
        self._float_sizes_timer.setInterval(_DOCKED_SIZES_DELAY_MS)
        self._float_sizes_timer.timeout.connect(self._persist_docked_sizes)
        self.content_splitter.splitterMoved.connect(self._on_splitter_moved)
        for key in self._floatable:
            self.float_controller.restore_saved(key)

        self.target_panel.target_changed.connect(self._on_target_changed)
        self.target_panel.scheme_changed.connect(self._on_scheme_changed)
        self.boundary_table.boundary_activated.connect(self._on_boundary_activated)
        self.scheme_summary.save_requested.connect(self.save_scheme)

    def _make_floatable(self, key: str, panel: QWidget, title: str) -> None:
        """Register a side panel for float/dock and give it a float button."""
        dock_manager.register_panel(key, title)
        self._floatable[key] = panel
        PanelFloatButton(key, panel, self.float_controller)

    def _on_splitter_moved(self, *_pos: int) -> None:
        self._float_sizes_timer.start()

    def _persist_docked_sizes(self) -> None:
        """Persist the splitter distribution for every docked side panel."""
        sizes = list(self.content_splitter.sizes())
        for key, panel in self._floatable.items():
            if panel.parentWidget() is self.content_splitter:
                self._float_persistence.save_docked_sizes(key, sizes)

    def set_project(self, project) -> None:
        self._project = project

    def update_state(self, stratigraphy) -> None:
        self.target_panel.update_state(stratigraphy)
        self.boundary_table.update_state(stratigraphy)
        self.scheme_summary.update_state(stratigraphy)

    def save_scheme(self) -> bool:
        """Persist current panel values into project.stratigraphy and bind downstream."""
        if self._project is None:
            QMessageBox.warning(self, "层序方案", "未绑定工程，无法保存")
            return False
        target = self.target_panel.current_target()
        scheme = self.target_panel.current_scheme()
        if not target:
            QMessageBox.warning(self, "层序方案", "请先选择或输入目标层位")
            return False
        apply_stratigraphy_scheme(
            self._project,
            target_horizon=target,
            systems_tract_scheme=scheme,
            bind_downstream=True,
        )
        self.update_state(self._project.stratigraphy)
        self.scheme_summary.set_bind_status(f"已保存 · 目标 {target}")
        self.stratigraphy_updated.emit()
        return True

    def _on_target_changed(self, target: str) -> None:
        if self._project is None or not target:
            return
        apply_stratigraphy_scheme(
            self._project,
            target_horizon=target,
            bind_downstream=True,
        )
        self.boundary_table.update_state(self._project.stratigraphy)
        self.scheme_summary.update_state(self._project.stratigraphy)
        self.scheme_summary.set_bind_status(f"已绑定 · 目标 {target}")
        self.stratigraphy_updated.emit()

    def _on_scheme_changed(self, scheme: str) -> None:
        if self._project is None or not scheme:
            return
        apply_stratigraphy_scheme(
            self._project,
            systems_tract_scheme=scheme,
            bind_downstream=False,
        )
        self.scheme_summary.update_state(self._project.stratigraphy)
        self.stratigraphy_updated.emit()

    def _on_boundary_activated(self, boundary: str) -> None:
        if self._project is None:
            return
        set_target_from_boundary(self._project, boundary, bind_downstream=True)
        self.update_state(self._project.stratigraphy)
        self.scheme_summary.set_bind_status(f"已绑定 · 目标 {boundary}")
        self.stratigraphy_updated.emit()

File: libs/ui_seqviz/src/qt/sequence_framework_page.cpp
#include <pwb/ui_seqviz/qt/sequence_framework_page.hpp>

#include <QVBoxLayout>

#include <pwb/ui_seqviz/qt/panel_float_button.hpp>
#include <pwb/ui_seqviz/qt/sequence_panels.hpp>
#include <pwb/ui_shell/dock_manager.hpp>
#include <pwb/ui_shell/float_controller.hpp>

namespace pwb::ui_seqviz::qt {

namespace {
constexpr const char* kTargetPanelKey = "sequence-target-panel";
constexpr const char* kSummaryPanelKey = "sequence-summary-panel";
}  // namespace

SequenceFrameworkPage::SequenceFrameworkPage(QWidget* parent)
    : QWidget(parent) {
    setObjectName("SequenceFrameworkPage");
    auto* layout = new QVBoxLayout(this);
    layout->setContentsMargins(0, 0, 0, 0);
    layout->setSpacing(0);

    splitter_ = new QSplitter(Qt::Horizontal, this);
    layout->addWidget(splitter_);

    target_panel_ = new SequenceTargetPanel(this);
    splitter_->addWidget(target_panel_);
    boundary_table_ = new SequenceBoundaryTable(this);
    splitter_->addWidget(boundary_table_);
    scheme_summary_ = new SequenceSchemeSummary(this);
    splitter_->addWidget(scheme_summary_);

    splitter_->setStretchFactor(0, 0);
    splitter_->setStretchFactor(1, 1);
    splitter_->setStretchFactor(2, 0);
    splitter_->setSizes({280, 900, 260});

    connect(target_panel_, &SequenceTargetPanel::target_changed, this,
            &SequenceFrameworkPage::on_target_changed);
    connect(target_panel_, &SequenceTargetPanel::scheme_changed, this,
            &SequenceFrameworkPage::on_scheme_changed);
    connect(boundary_table_, &SequenceBoundaryTable::boundary_activated,
            this, &SequenceFrameworkPage::on_boundary_activated);
    connect(scheme_summary_, &SequenceSchemeSummary::save_requested, this,
            [this] { save_scheme(); });
}

SequenceFrameworkPage::~SequenceFrameworkPage() = default;

const StratigraphySlice& SequenceFrameworkPage::stratigraphy() const {
    static const StratigraphySlice kEmpty;
    return project_ != nullptr ? project_->stratigraphy : kEmpty;
}

void SequenceFrameworkPage::set_project(
    StratigraphyProjectSlice* project) {
    project_ = project;
}

void SequenceFrameworkPage::set_float_controller(
    ui_shell::LayoutPersistence* persistence) {
    ui_shell::dock_manager().register_panel(kTargetPanelKey, "层序格架设置");
    ui_shell::dock_manager().register_panel(kSummaryPanelKey, "层序方案摘要");

    auto resolver = [this](const std::string& key) -> QWidget* {
        if (key == kTargetPanelKey) {
            return target_panel_;
        }
        if (key == kSummaryPanelKey) {
            return scheme_summary_;
        }
        return nullptr;
    };
    auto title_for = [](const std::string& key) -> QString {
        return key == kTargetPanelKey
                   ? QStringLiteral("层序格架设置")
                   : QStringLiteral("层序方案摘要");
    };
    float_controller_ = std::make_unique<ui_shell::FloatController>(
        std::move(resolver), persistence, std::move(title_for), this);

    // Owned by their panels; kept alive by the panel parent chain.
    new PanelFloatButton(kTargetPanelKey, target_panel_,
                         float_controller_.get());
    new PanelFloatButton(kSummaryPanelKey, scheme_summary_,
                         float_controller_.get());
    float_controller_->restore_saved(kTargetPanelKey);
    float_controller_->restore_saved(kSummaryPanelKey);
}

void SequenceFrameworkPage::update_state(
    const StratigraphySlice& stratigraphy) {
    target_panel_->update_state(stratigraphy);
    boundary_table_->update_state(stratigraphy);
    scheme_summary_->update_state(stratigraphy);
}

bool SequenceFrameworkPage::save_scheme() {
    if (project_ == nullptr) {
        return false;
    }
    const QString scheme = target_panel_->current_scheme();
    if (scheme.isEmpty()) {
        emit warning_requested(QStringLiteral("缺少体系域方案"),
                               QStringLiteral("请输入体系域方案"));
        return false;
    }
    const QString target = target_panel_->current_target();
    const StratigraphySlice& strat = apply_stratigraphy_scheme(
        *project_, target.isEmpty() ? std::nullopt
                                    : std::optional<std::string>(
                                          target.toStdString()),
        scheme.toStdString(), std::nullopt, std::nullopt,
        /*bind_downstream=*/true);
    update_state(strat);
    scheme_summary_->set_bind_status(
        QStringLiteral("已保存到项目（%1）").arg(scheme));
    emit stratigraphy_updated();
    return true;
}

void SequenceFrameworkPage::on_target_changed(const QString& target) {
    if (project_ == nullptr || target.trimmed().isEmpty()) {
        return;
    }
    const StratigraphySlice& strat = apply_stratigraphy_scheme(
        *project_, target.trimmed().toStdString(), std::nullopt,
        std::nullopt, std::nullopt, /*bind_downstream=*/true);
    update_state(strat);
    scheme_summary_->set_bind_status(
        QStringLiteral("已应用到项目（%1）").arg(target.trimmed()));
    emit stratigraphy_updated();
}

void SequenceFrameworkPage::on_scheme_changed(const QString& scheme) {
    if (project_ == nullptr || scheme.trimmed().isEmpty()) {
        return;
    }
    const StratigraphySlice& strat = apply_stratigraphy_scheme(
        *project_, std::nullopt, scheme.trimmed().toStdString(),
        std::nullopt, std::nullopt, /*bind_downstream=*/true);
    update_state(strat);
    scheme_summary_->set_bind_status(
        QStringLiteral("已应用到项目（%1）").arg(scheme.trimmed()));
    emit stratigraphy_updated();
}

void SequenceFrameworkPage::on_boundary_activated(const QString& name) {
    if (project_ == nullptr || name.trimmed().isEmpty()) {
        return;
    }
    const StratigraphySlice& strat = set_target_from_boundary(
        *project_, name.toStdString(), /*bind_downstream=*/true);
    update_state(strat);
    scheme_summary_->set_bind_status(
        QStringLiteral("已设为目标层位（%1）").arg(name.trimmed()));
    emit stratigraphy_updated();
}

}  // namespace pwb::ui_seqviz::qt
